### ISSUE TITLE
[Dart 3.0] GetIt can't detect T type properly

### ISSUE THREAD

**Issue by subzero911** [fluttercommunity/get_it#331]
After the latest Flutter upgrade, registerSingleton inside of VoidCallback is not working anymore:

<img width="664" alt="image" src="https://github.com/fluttercommunity/get_it/assets/12999702/fd092d77-4210-4010-9447-99bb2401a80f">

It worked before Dart 3.0. Now they probably have stricter type checking.
In runtime it crashes with the following error
![image](https://github.com/fluttercommunity/get_it/assets/12999702/cf09e074-f380-4877-bc4e-81f0f3f97baa)

The bug disappears when I explicitly indicate the type:
![image](https://github.com/fluttercommunity/get_it/assets/12999702/4e9f3ad2-9e19-4730-99b5-7cc34315e43a)

This bug appears in VoidCallback only, not in a plain code. It also can be fixed if I used:
`() {}` instead of `() =>`

<img width="437" alt="image" src="https://github.com/fluttercommunity/get_it/assets/12999702/4392514b-82e3-417e-8bc0-6a2e48691ca4">

The reason behind: a generic by default takes the root return type (void), and Dart tries to put `void` into `T extends Object` constraint, and this apparently causes this assert to be falsely posivite:
![image](https://github.com/fluttercommunity/get_it/assets/12999702/0d374fb8-d086-41b5-b11d-2e6c365d169f)
 
Any ideas on it?

**Comment by escamoteur** [comment]
@subzero911 your analysis seems correct, although I don't think that this can be fixed by GetIt. I always recommend passing the Generic type explicitly because then there is no room for doubt

@mraleph @munificent this comes sort of surprising, is that the intended consequence?

**Comment by subzero911** [comment]
First of all, it's strange that we can do call  GetIt.instance(). Is it a callable class? I personally never did it. Do we actually need this assertion?

**Comment by escamoteur** [comment]
Actually the assertion isn't about calling it this way, but if people call GetIt.instance() accidentially you get this type of eception too.
And yes GetIt  is a callable class from the beginning.
I most of the time to something like this in my setup file:

´´´dart
final di = GetIt.instance;
// and then use it like

**Comment by subzero911** [comment]
> I always recommend passing the Generic type explicitly because then there is no room for doubt

But writing `GetIt.I.registerSingleton<MyControllerWithVeryLongName>(MyControllerWithVeryLongName())` every time is not concise.

**Comment by escamoteur** [comment]
That is true, but you don't register that many types and in my case I often register an abstract interface class with a selected implementation then I have to provide the type anyway
Am 29. Mai 2023, 11:21 +0200 schrieb Sergey Molchanovsky ***@***.***>:

**Comment by munificent** [comment]
I'm not sure why this behavior is different in Dart 3.0. @stereotype441 or @eernst might know more.

**Comment by escamoteur** [comment]
@mraleph but why did that change with Dart 3.0 and was working before?

**Comment by mraleph** [comment]
> I mean if you look at the code example at the top, although I always recommended to provide the generic type, registerSingleton was working perfectly infering the type from the passed instance.
> only when called in a void handler, it fails

Yep. You can see this in the reduced example I have provided above and I think it is working as intended and has been like that for a very long time. Dart's inference is known to sometimes give too much attention to the context type rather than looking at parameter types. 

My guess, as I said, would be that inference _somewhere else_ got more precise and as a result `void` got inferred as a context type where it was not before. But hard to confirm this without having full code. 

Just to make it more obvious:

```dart
void foo<R>(R Function() f) {
  f();
}

T bar<T extends Object>(T v) {
  print(T);
  return v;
}

void main() {
  foo<void>(() => bar(''));  // => Object
  foo<void>(() { bar(''); });  // => String
  foo<dynamic>(() => bar(''));  // => String
}
```
